ethcontract-common/src/artifact/hardhat.rs

```diff
@@ -139,42 +156,158 @@ impl HardHatLoader {
-        self.load_artifact("<unknown>", v, from_reader, from_reader)
+    /// Add contract name to the list of [`allowed contracts`].
+    ///
+    /// [`allowed contracts`]: #structfield.contracts_allow_list
+    pub fn allow_contract(mut self, contract: impl Into<String>) -> Self {
+        self.contracts_allow_list.push(contract.into());
+        self
+    }
+
+    /// Add contract name to the list of [`denied contracts`].
+    ///
+    /// [`denied contracts`]: #structfield.contracts_deny_list
+    pub fn deny_contract(mut self, contract: impl Into<String>) -> Self {
+        self.contracts_deny_list.push(contract.into());
+        self
     }
 
-    /// Loads an artifact from bytes of JSON text.
-    pub fn load_from_slice(&self, v: &[u8]) -> Result<Artifact, ArtifactError> {
-        self.load_artifact("<unknown>", v, from_slice, from_slice)
+    /// Load an artifact from a JSON value.
+    pub fn load_from_reader(&self, f: Format, v: impl Read) -> Result<Artifact, ArtifactError> {
+        self.load_artifact(f, "<unknown>", v, from_reader, from_reader)
     }
 
-    /// Loads an artifact from string of JSON text.
-    pub fn load_from_str(&self, v: &str) -> Result<Artifact, ArtifactError> {
-        self.load_artifact("<unknown>", v, from_str, from_str)
+    /// Load an artifact from bytes of JSON text.
+    pub fn load_from_slice(&self, f: Format, v: &[u8]) -> Result<Artifact, ArtifactError> {
+        self.load_artifact(f, "<unknown>", v, from_slice, from_slice)
     }
 
-    /// Loads an artifact from a loaded JSON value.
-    pub fn load_from_value(&self, v: Value) -> Result<Artifact, ArtifactError> {
-        self.load_artifact("<unknown>", v, from_value, from_value)
+    /// Load an artifact from string of JSON text.
+    pub fn load_from_str(&self, f: Format, v: &str) -> Result<Artifact, ArtifactError> {
+        self.load_artifact(f, "<unknown>", v, from_str, from_str)
     }
 
-    /// Loads an artifact from disk.
-    pub fn load_from_file(&self, p: impl AsRef<Path>) -> Result<Artifact, ArtifactError> {
+    /// Load an artifact from a loaded JSON value.
+    pub fn load_from_value(&self, f: Format, v: Value) -> Result<Artifact, ArtifactError> {
+        self.load_artifact(f, "<unknown>", v, from_value, from_value)
+    }
+
+    /// Load an artifact from disk.
+    pub fn load_from_file(
+        &self,
+        f: Format,
+        p: impl AsRef<Path>,
+    ) -> Result<Artifact, ArtifactError> {
         let path = p.as_ref();
         let file = File::open(path)?;
         let reader = BufReader::new(file);
-        self.load_artifact(path.display(), reader, from_reader, from_reader)
+        self.load_artifact(f, path.display(), reader, from_reader, from_reader)
+    }
+
+    /// Load an artifact from `deployments` directory.
+    pub fn load_from_directory(&self, p: impl AsRef<Path>) -> Result<Artifact, ArtifactError> {
+        self._load_from_directory(p.as_ref())
```

Comment:
Is there some added benefit to adding a helper method with the same name for just this one line?

Reply:
This is done to avoid code bloat. `impl AsRef<Path>` makes this method generic, so its body will be monomorphized multiple times depending on actual type of the argument. Also, the compiler will not emit assembly for generic functions when compiling them. Instead, it will save their IR and turn it to assembly whenever it sees the function used. This means that generics will get compiled multiple times, once per crate per argument type.

Reply:
https://gnosisinc.slack.com/archives/C01PBR5TWCS/p1624448800012900

Reply:
The reason is to make it more ergonomic for callers so that they don't have to use `as_ref()`. We had a discussion about this on slack in the rust channel a few weeks ago.